sidecar/src/interface.rs

```diff
@@ -404,6 +436,65 @@ impl EnqueuedTelemetryData {
             actions.push(TelemetryActions::AddIntegration(i.clone()));
         }
     }
+
+    pub async fn process_immediately(sidecar_actions: Vec<SidecarAction>) -> Vec<TelemetryActions> {
+        let mut actions = vec![];
+        for action in sidecar_actions {
+            match action {
+                SidecarAction::Telemetry(t) => actions.push(t),
+                SidecarAction::PhpComposerTelemetryFile(path) => {
+                    for nested in Self::extract_composer_telemetry(path).await.iter() {
+                        actions.push(TelemetryActions::AddDependecy(nested.clone()));
+                    }
+                }
+            }
+        }
+        actions
+    }
+
+    // This parses a vendor/composer/installed.json file. It caches the parsed result for a while.
+    fn extract_composer_telemetry(path: PathBuf) -> ManualFuture<Arc<Vec<data::Dependency>>> {
+        let (deps, completer) = ManualFuture::new();
+        tokio::spawn(async {
+            type ComposerCache = HashMap<PathBuf, (SystemTime, Arc<Vec<data::Dependency>>)>;
+            lazy_static! {
+                static ref COMPOSER_CACHE: tokio::sync::Mutex<ComposerCache> =
+                    tokio::sync::Mutex::new(Default::default());
+            }
+
+            let mut cache = COMPOSER_CACHE.lock().await;
+            let packages = match tokio::fs::metadata(&path).await.and_then(|m| m.modified()) {
+                Err(e) => {
+                    warn!("Failed to report dependencies from {path:?}, could not read modification time: {e:?}");
+                    Arc::new(vec![])
+                }
+                Ok(modification) => {
+                    let now = SystemTime::now();
+                    if let Some((last_update, actions)) = cache.get(&path) {
+                        if modification < *last_update {
+                            completer.complete(actions.clone()).await;
+                            return;
+                        }
+                    }
+                    async fn parse(path: &PathBuf) -> Result<Vec<data::Dependency>> {
+                        let mut json = tokio::fs::read(&path).await?;
+                        let parsed: ComposerPackages = simd_json::from_slice(json.as_mut_slice())?;
+                        Ok(parsed.packages)
+                    }
+                    let packages = Arc::new(parse(&path).await.unwrap_or_else(|e| {
+                        warn!("Failed to report dependencies from {path:?}: {e:?}");
+                        vec![]
+                    }));
+                    cache.insert(path, (now, packages.clone()));
+                    // cheap way to avoid unbounded caching
+                    cache.retain(|_, (inserted, _)| *inserted > now.sub(Duration::from_secs(2000)));
```

Comment:
So retain is `O(capacity)` since it visits all buckets (even empty ones), but I guess that there aren't that many entries (i.e. capacity is small) and/or not that many updates to the files.